File: crates/dewasm-backend-go/units/wasi/fd_prestat_dir_name.go
// requires: memory/init, wasi/errno_fs
func (w *WASI) wasi_fd_prestat_dir_name(fd, pathPtr, pathLen uint32) uint32 {
    d, ok := w.fds[fd].(*wasiDir)
    if !ok || d.preopenName == nil {
        return wasiBadf
    }
    name := d.preopenName
    if uint32(len(name)) > pathLen {
        return wasiNametoolong
    }
    w.memory.init(uint64(pathPtr), name, 0, uint64(len(name)))
    return wasiOk
}

File: crates/dewasm-backend-java/units/wasi/fd_prestat_dir_name.java
// requires: memory/init, wasi/errno_fs
int wasi_fd_prestat_dir_name(int fd, int pathPtr, int pathLen) {
    Object e = fds.get(fd);
    if (!(e instanceof Dir) || ((Dir) e).preopenName == null) {
        return WASI_BADF;
    }
    byte[] name = ((Dir) e).preopenName;
    if (name.length > Integer.toUnsignedLong(pathLen)) {
        return WASI_NAMETOOLONG;
    }
    memory.init(Integer.toUnsignedLong(pathPtr), name, 0, name.length);
    return WASI_OK;
}
